### PR TITLE
Development/vfs redesign

### PR DESCRIPTION
**Pull request by RisingLiberty**
This is to resolve https://github.com/New-Creta/Rex/issues/238
Removed 2 bool flags and using an enum instead.
Also removed the thread spawning when waiting for tasks to complete

### CODE REVIEW

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+        {
+          // get the queued request and remvoe it from the queue,
+          // but remove in the reverse order that they got added
+          rsl::string_view filepath = g_read_requests_in_order.front();
+          g_read_requests_in_order.erase(g_read_requests_in_order.cbegin());
+          rsl::unique_ptr<QueuedRequest> request = rsl::move(g_read_requests[filepath]);
+          g_read_requests.erase(filepath);
+
+          // we don't need access to the queue anymore, we can unlock its access mutex
+          lock.unlock();
```
Comment (Dyronix) [inline source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:330]: NIT: I always add another scope to make sure the dtor is triggered, if this code ever changes it's done automatically without me thinking about the unlock.

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp [hunk 2]
```diff
@@ -0,0 +1,637 @@
...
+      // This thread processes the read requests
+      // A queued request is popped from the queue
+      // reads the file specified as the requested file to read
+      // signals all the read requests that this file is read
+      // and passed in the buffer with its size
+      // After which it waits async for all read requests to have finished processing
+      // the data passed in to them
+      // Finally it passes itself over to the closing thread queue
+      // to make sure the data stays alive until all requests have finished
+      g_reading_thread = rsl::thread(process_read_requests);
```
Comment (Dyronix) [inline source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:390]: INFO: This is probably already planned for the future, but wouldn't a job system be a better pick here? just spawning a new "job" that handles these read request. Would also make it possible to read multiple files at once.

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+
+      // Start the worker threads, listening to file IO requests and processing them
+      start_threads();
+
+      g_vfs_state_controller.change_state(VfsState::Running);
+
+      REX_LOG(FileSystem, "FileSystem initialized");
+    }
+
+    void set_root(rsl::string_view root)
```
Comment (Dyronix) [inline source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:413]: INFO: Are we ever required to call this function?
I think we as engine-developers control where the root of the VFS points to and users of the engine should just go with it no?

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+        return rsl::string(path);
+      }
+
+      const auto comparer = rsl::equal_to<rsl::string>();
+      using K1 = rsl::string;
+      using K2 = rsl::string_view;
+      using Key = rsl::string;
+
+      K1 k1;
+      K2 k2;
```
Comment (Dyronix) [inline source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:628]: INFO: Why is this here?

**Review by Dyronix — CHANGES_REQUESTED** [review]

File: source/1_engine/rex_engine/src/filesystem/path.cpp [hunk 5]
```diff
@@ -81,6 +81,16 @@ namespace rex
...
+    // removes leading and trailing quotes from a path
+    rsl::string_view remove_quotes(rsl::string_view path)
+    {
+      if (path.starts_with("\"") && path.ends_with("\""))
+      {
+        path = path.substr(1, path.length() - 2);
+      }
+
+      return path;
+    }
```
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/filesystem/path.cpp:93]: true, but it's in the path namespace, where else would you expect it to remove quotes?

File: source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h [hunk 6]
```diff
@@ -8,6 +8,40 @@
...
+/// File Reading:
+// rex::memory::Blob content = rex::vfs::read_file("path/to/file");
+///  
+/// File Writing:
+// int x = 0;
+// rex::vfs::write_to_file("path/to/file.txt", &x, sizeof(x);
+///
+/// ASYNC FILE IO
+/// 
+/// We only support async file reading, as async file writing
```
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h:27]: part of it is one the next line, the full sentence is: We only support async file reading, as async file writing
/// Doesn't make much sense for the moment

File: source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h [hunk 7]
```diff
@@ -6,5 +6,5 @@ namespace rex
 }
 
 #ifdef REX_PLATFORM_WINDOWS
-#include "rex_engine/platform/win/filesystem/vfs.h"
+#include "rex_engine/platform/win/filesystem/win_vfs.h"
```
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h:9]: no, I made it win_vfs to make it easier on intellisense and so you can see in the IDE which file you're editing, the windows specific version of the cross platform version

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp [hunk 8]
```diff
@@ -0,0 +1,637 @@
...
+      m_queued_request = other.m_queued_request;
+      m_is_done        = other.m_is_done;
+      m_buffer         = other.m_buffer;
+      m_size           = other.m_size;
+
+      m_queued_request->add_request_to_signal(this);
+
+      return *this;
+    }
+    ReadRequest& ReadRequest::operator=(ReadRequest&& other)
```
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:243]: not needed

File: source/1_engine/rex_engine/src/filesystem/path.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Dyronix) [reply source/1_engine/rex_engine/src/filesystem/path.cpp:93]: Well, I could have made a typo and have two quotes at the start and one at the end it will not capture that. 
Or I could have quotes within my path, I know this is not a valid path, but people make mistakes.

Just getting ahead of problems that might occur.

File: source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dyronix) [reply source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h:27]: But we don't "only" support async. 
We also support sync, does are the lines of code above or am I mistaken?

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp (hunk 3 above)
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:413]: this is more for debugging stuff (eg auto tests)

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp (hunk 4 above)
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:628]: left over debug code

File: source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dyronix) [reply source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h:9]: This is different from the other platform specific code we are using, I would pick one and go with it. Please do not mix and match.

I can see it in my IDE which one I am editing if I look at the Solution Explorer (in VS) or the the file browser (in Code)

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Dyronix) [reply source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:243]: I add this keyword to all the "move" ctors or move assignment operators are you telling me this is redundant?
Is it because we don't have expections enabled?

File: source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h:27]: the "only" isn't on the "async" part, it's on the "file reading" part, we only support async file read operations, we don't support async file writing operations

File: source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h:9]: these mistakes of not seeing which file you're operating in get more obvious when rex gets bigger. yes you can see it in solution explorer or in the browser, but those things can easily be overlooked.
also swapping between header and cpp file can get confusing if the files have the same name
the format we should be using is, if there's a cross platform filename for something, you add a prefix of the platform to the file in the platform specific code. if there's no cross platform equivalent, the platform prefix isn't needed.

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp:243]: yes

File: source/1_engine/rex_engine/src/filesystem/path.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/filesystem/path.cpp:93]: if you want to remove quotes in the middle, you can't return a string view, it'll have to get reverted to a string.

[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/filesystem/path.cpp:93]: regarding a mis match in the number of quotes at the start and the end, yes that's fixable, but that'd change the behavior of this function. I've had that before and had to change it because it caused an issue, although I don't remember what caused it

[the remark this answers is not in the document]
Comment (Dyronix) [reply source/1_engine/rex_engine/src/filesystem/path.cpp:93]: But the most important part is that I would make it clear that it will only remove if start AND end have quotes. if you only have quotes at the start it won't do anything.

File: source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (Dyronix) [reply source/1_engine/rex_engine/include/rex_engine/platform/win/filesystem/win_vfs.h:27]: Maybe there needs to be a change to this comment here as I got confused twice.

File: source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dyronix) [reply source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h:9]: I don't understand why you suddenly change again, you did this for the paths but for the VFS you are reluctant again. Either change the path ( and everything else that uses this tactic to this strategy ) or use the design you used for the path on this thing as well but please don't mix and match it will get confusing very fast.

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp (hunk 1 above)
Comment (Dyronix) [reply source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp]: You added a thump here but I don't see this change? 
Was this just to verify you read it?

File: source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (RisingLiberty) [reply source/1_engine/rex_engine/include/rex_engine/filesystem/vfs.h:9]: it's possible I forgot something, I thought I changed path and vfs to be win_path and win_vfs, it's possible due to all the merges, some of that got lost

File: source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp (hunk 1 above)
Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/platform/win/filesystem/win_vfs.cpp]: I put an extra scope around the entire if block, it's impossible for me to start before the if and end halfway through

**Review by Dyronix — APPROVED** [review]

**Review by Dyronix — DISMISSED** [review]

**Review by Dyronix — APPROVED** [review]
